### PR TITLE
HW6

### PR DESCRIPTION
**Pull request by DataJoshua**
### Summary

A brief description of the pull request.

### How it works

Screenshots/screencasts of the pull request introduced functionality.

### Test plan

List of steps to manually test introduced functionality:

- Log in
- Go to http://localhost:3000

### Deploy notes

Notes regarding deployment the contained body of work.
These should note any db migrations, services, scripts, etc.

### Checklist:

- [ ] I have performed a self-review of my own code
- [ ] My pull request is shorter than 500 lines

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
+class CommentsController < ApplicationController
+  before_action :set_project, only: %i[index new create edit update destroy] 
+  before_action :set_task, only: %i[index show edit destroy new create update]
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:3]: You don't need `only` if there is all methods

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  def show
+  end
+
+  def new
+    @comment = Comment.new
+  end
+
+  def create
+    @comment = Comment.new(comment_params.merge(user_id: current_user.id, task_id: @task.id))
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:18]:
```suggestion
    @comment = Comment.new(comment_params.merge(user: current_user, task: @task))
```

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    if @comment.save 
+      redirect_to project_task_comments_path(@project, @task)
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:21]: Does it work?

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  def index
+    @comments = @task.comments
+  end
+
+  def show
+  end
+
+  def new
+    @comment = Comment.new
+  end
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:15]: We don't need this actions

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      redirect_to [@project, @task, @comment]
+    else
+      render :edit
+    end
+  end
+
+  def destroy
+    @comment.destroy
+    flash[:notice] = "Your comment has been deleted"
+    redirect_to project_task_comments_path(@project, @task)
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:40]: Does it work?

File: app/controllers/comments_controller.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def set_project
+    @project = Project.find(params[:project_id])
+  end
+
+  def comment_params
+    params.require(:comment).permit(:content)
+  end
+
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:60]: Redundant empty line

File: app/controllers/concerns/authorization.rb [hunk 7]
```diff
@@ -2,7 +2,6 @@ module Authorization
   extend ActiveSupport::Concern
 
   included do
-    verify_authorized
```
Comment (RasimKhusaenov) [inline app/controllers/concerns/authorization.rb:5]: Why?

File: app/models/task.rb [hunk 8]
```diff
@@ -1,6 +1,6 @@
 class Task < ApplicationRecord
   belongs_to :project
-
+  has_many :comments
```
Comment (RasimKhusaenov) [inline app/models/task.rb:3]: https://docs.rubocop.org/rubocop-rails/cops_rails.html#railshasmanyorhasonedependent

File: app/models/user.rb [hunk 9]
```diff
@@ -7,7 +7,8 @@ class User < ApplicationRecord
 
   has_many :project_memberships, dependent: :restrict_with_error
   has_many :projects, through: :project_memberships
-
+  has_many :comments
```
Comment (RasimKhusaenov) [inline app/models/user.rb:10]: https://docs.rubocop.org/rubocop-rails/cops_rails.html#railshasmanyorhasonedependent

File: app/views/comments/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<h1>Edit Comment</h1>
+
+<%= form_with model: [@project, @task, @comment] do |form| %>
```
Comment (RasimKhusaenov) [inline app/views/comments/edit.html.erb:3]: Very good

Comment (RasimKhusaenov) [reply app/views/comments/edit.html.erb:3]: Can you move form to partial?

File: app/views/comments/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+<h1>Comments#new</h1>
+<p>Find me in app/views/comments/new.html.erb</p>
```
Comment (RasimKhusaenov) [inline app/views/comments/new.html.erb:2]: User doesn't should know about files structure

File: config/database.yml [hunk 12]
```diff
@@ -22,8 +22,13 @@ default: &default
...
 development:
-  <<: *default
-  database: task_tracker_development
+  adapter: postgresql
+  encoding: unicode
+  database: taskTracker
+  host: localhost
+  pool: 5
+  username: postgres
+  password: MyPassword
```
Comment (RasimKhusaenov) [inline config/database.yml:31]: Don't change this file

File: config/routes.rb [hunk 13]
```diff
@@ -2,7 +2,9 @@
   root to: "projects#index"
 
   resources :projects do
-    resources :tasks
+    resources :tasks do
+      resources :comments
```
Comment (RasimKhusaenov) [inline config/routes.rb:6]: We don't need all pages

File: db/migrate/20221111133354_add_column_state_to_tasks.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+class AddColumnStateToTasks < ActiveRecord::Migration[6.1]
+  def change
+    add_column :tasks, :state, :string, default: "Not Started"
```
Comment (RasimKhusaenov) [inline db/migrate/20221111133354_add_column_state_to_tasks.rb:3]: Where is it in `app/models/task.rb`?

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

**Comment by RasimKhusaenov** [comment]
Where is policies?

File: app/views/comments/edit.html.erb (hunk 10 above)
Comment (DataJoshua) [reply app/views/comments/edit.html.erb]: yea )

File: app/policies/comment_policy.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
+class CommentPolicy < ApplicationPolicy
+  authorize :user, allow_nil: true
```
Comment (RasimKhusaenov) [inline app/policies/comment_policy.rb:2]: Why we need it?

File: app/policies/session_policy.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+  def show?
+    user.present?
+  end
+
+  def new?
+    create?
+  end
+
+  def create?
+    !user.present?
```
Comment (RasimKhusaenov) [inline app/policies/session_policy.rb:13]: Use `blank?` instead

File: app/policies/task_policy.rb [hunk 17]
```diff
@@ -0,0 +1,37 @@
+class TaskPolicy < ApplicationPolicy
+  authorize :user, optional: true
```
Comment (RasimKhusaenov) [inline app/policies/task_policy.rb:2]: Why we need it?

File: app/controllers/projects_controller.rb [hunk 18]
```diff
@@ -21,6 +21,7 @@ def new
 
   # GET /projects/1/edit
   def edit
+    authorize! @project
```
Comment (RasimKhusaenov) [inline app/controllers/projects_controller.rb:24]: Can you move it to `before_action`?

File: app/controllers/comments_controller.rb [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  before_action -> { authorize! @task }
+
+  def new
+    @comment = Comment.new
+  end
+
+  def create
+    @comment = Comment.new(comment_params.merge(user: current_user, task: @task))
+
+    if @comment.save
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:14]: else?

File: app/controllers/comments_controller.rb [hunk 20]
```diff
@@ -0,0 +1,55 @@
+class CommentsController < ApplicationController
+  before_action :set_project
+  before_action :set_task
+  before_action :set_comment, only: %i[edit update destroy]
+  before_action -> { authorize! @task }
+
+  def new
+    @comment = Comment.new
+  end
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:9]: We don't need it

File: app/views/tasks/show.html.erb [hunk 21]
```diff
@@ -25,3 +25,16 @@
...
+
+<% @task.comments.each do |comment| %>
+  <h4> User #<%= comment.user.first_name %> </h4>
+  <p> <%= comment.content %> </p>
+  <%= link_to "edit", edit_project_task_comment_path(@task.project, @task, comment) %>
+  <%= link_to "delete", project_task_comment_path(@task.project, @task, comment), method: :delete %> 
+  <hr>
+<% end %>
+
+<%= link_to 'New Comment', new_project_task_comment_path(@task.project, @task) %>
```
Comment (RasimKhusaenov) [inline app/views/tasks/show.html.erb:40]: You should create comment inside this page, not in `new`

File: app/policies/comment_policy.rb (hunk 15 above)
Comment (RasimKhusaenov) [reply app/policies/comment_policy.rb:2]: This line allows the current user to be empty. The default value for "user" does not depend on this.

Comment (DataJoshua) [reply app/policies/comment_policy.rb:2]: so we need the user to be always present?

Comment (RasimKhusaenov) [reply app/policies/comment_policy.rb:2]: For our case yes

File: app/policies/comment_policy.rb [hunk 22]
```diff
@@ -1,5 +1,5 @@
 class CommentPolicy < ApplicationPolicy
-  authorize :user, allow_nil: true
+  authorize :user
```
Comment (RasimKhusaenov) [inline app/policies/comment_policy.rb:2]: It is redundant. https://actionpolicy.evilmartians.io/#/custom_policy?id=custom-base-policy
```
  # ActionPolicy::Policy::Defaults module adds the following

  authorize :user
```

File: app/controllers/comments_controller.rb [hunk 23]
```diff
@@ -0,0 +1,56 @@
+class CommentsController < ApplicationController
+  before_action :set_project
+  before_action :set_task
+  before_action :set_comment, only: %i[edit update destroy]
+  before_action -> { authorize! @task }
+
+  def new
+  end
```
Comment (RasimKhusaenov) [inline app/controllers/comments_controller.rb:8]: WE DON'T NEED IT AT ALL

Comment (RasimKhusaenov) [reply app/controllers/comments_controller.rb:8]: You don't have route for this action, it is redundant. Do you understand it?

File: config/routes.rb [hunk 24]
```diff
@@ -2,7 +2,9 @@
   root to: "projects#index"
 
   resources :projects do
-    resources :tasks
+    resources :tasks do
+      resources :comments, only: %i[new create edit update destroy]
```
Comment (RasimKhusaenov) [inline config/routes.rb:6]: Why you need `new`? You should create comment in task page.

File: app/controllers/projects_controller.rb [hunk 25]
```diff
@@ -46,6 +47,7 @@ def update
 
   # DELETE /projects/1
   def destroy
+    authorize! @project
```
Comment (RasimKhusaenov) [inline app/controllers/projects_controller.rb:50]: Move it too

File: app/controllers/comments_controller.rb (hunk 23 above)
Comment (DataJoshua) [reply app/controllers/comments_controller.rb:8]: I got it

**Pull request closed without merging**